paxml/trainer_lib.py

```diff
@@ -1130,8 +1130,15 @@ def train_step_single_learner(
     wps_with_opt = tasks_lib.filter_vars_for_grad_or_opt(
         var_weight_hparams, excluded_for_opt
     )
+    # Exclude the overwrite_with_gradient params, since they don't go through
+    # the optimizer and they don't have corresponding opt states.
+    overwrite_var_mask = \
+        tasks_lib.get_overwrite_with_gradient_var_mask(var_weight_hparams)
+    normal_grads = tasks_lib.filter_vars_for_grad_or_opt(
```

Comment:
the learner handels non-trainable variables by running this code

    var_is_learnable = jax.tree_util.tree_map(
        lambda x: not base_layer.var_not_trainable(x), var_weight_hparams
    )

We should either add another condition to the learner, or require all `overwrite_with_gradient` variables to be non-trainable. The latter makes sense to me. How about that?

Reply:
But do we need to pass the grads of 'overwrite_with_gradient' as part of `transformed_grads` and pass it to the apply_gradients(). So, something like:
```python
    transformed_grads, new_opt_states = learner.update_states(
        normal_grads, states.opt_states[0], vars_with_opt, wps_with_opt
    )
    transformed_grads = treemap(lambda a, b, c: b if a else c, overwrite_mask, transformed_grads, grads) 
    vars_with_opt = learner.apply_gradient(
        vars_with_opt, transformed_grads, wps_with_opt
    )
```
And inside the learners, I can add a branch to update the grads as you mentioned.

But I feel the above logic is more complicated, since we need to hijack the transformed_grads.

Reply:
The code you pasted is from the `apply_gradient()` of `learners.py` in paxml. But the `apply_gradient()` we used in [here](https://github.com/google/paxml/blob/main/paxml/trainer_lib.py#L1136) is from optax, since it doesn't accept the `var_weight_hparams`. So, unless we want to introduce `var_weight_hparams` to the apply_gradient() to optax optimizers, I don't think it is easy to put the logic of updating `overwrite_with_gradient` var inside learner.

Reply:
The above is wrong and `apply_gradient()` is indeed in the learners.py rather in optax.

However, I feel if we want to do the variable update inside the `apply_gradient()`, we would need to do lots of changes. Currently, the logic is like:
```
var_with_opt = ...
wps_with_opt = ...
grads_with_opt = ... # we added this
transformed_grads, new_opt_states = learner.update_states(grads_with_opt, states, vars_with_opt, wps_with_opt)
vars_with_opt = learner.apply_gradient(vars_with_opt, transformed_grads, wps_with_opt)
mdl_vars = vars_with_opt if has_grad else mdl_vars
mdl_vars = grads if is_overwrite else mdl_vars # we added this
```
Basically, we need to exclude the overwrite-with-grad params and at the very end, do the overwrite.

However, if we move the logic into the learner.apply_gradient(), the callsite of the above will be changed to:
```
var_with_opt = ...
wps_with_opt = ...
grads_with_opt = ... # we still add this
transformed_grads, new_opt_states = learner.update_states(grads_with_opt, states, vars_with_opt, wps_with_opt)
var_with_opt_and_overwrite = ... # we need to add this
transformed_grads = grads if is_overwrite else transformed_grads # we need to merge the grads
wps_with_opt_and_overwrite = ... # we need to add this
vars_with_opt = learner.apply_gradient(var_with_opt_and_overwrite, transformed_grads, wps_with_opt_and_overwrite)
mdl_vars = vars_with_opt if has_grad else mdl_vars
# mdl_vars = grads if is_overwrite else mdl_vars # Yes, we don't need this, however, we need to do more in apply_gradient()
```
Then, in the apply_gradient() we do:
```
exclude the overwrite from var_with_opt_and_overwrite, transformed_grads, wps_with_opt_and_overwrite # we need to add this
do the original updating
get the masks of overwrite for var_with_opt_and_overwrite, transformed_grads, wps_with_opt_and_overwrite # we need to add this
do the update for the overwrite # we need to add this
```
So, to summarize, I just feel the new solution make the code much complex. Or do I misunderstand your suggestion? @lukaszlew

Reply:
I think this is actually easier than I thought. I have pushed a new commit to move the updating of the overwrite-with-grads params into the apply_gradients. So, I think this comment can be addressed. (Sorry for the confusion.) @lukaszlew